File: templates/typescript/generator-bot-virtualassistant/generators/app/templates/sample-assistant/src/adapters/defaultAdapter.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import {
    BotFrameworkAdapter,
    BotFrameworkAdapterSettings,
    BotTelemetryClient,
    ConversationState,
    SetSpeakMiddleware,
    ShowTypingMiddleware,
    SkillHttpClient,
    TranscriptLoggerMiddleware,
    TranscriptStore,
    TurnContext,
    TelemetryException } from 'botbuilder';
import { BlobsTranscriptStore } from 'botbuilder-azure-blobs';
import {
    EventDebuggerMiddleware,
    LocaleTemplateManager,
    SetLocaleMiddleware } from 'bot-solutions';
import { TelemetryInitializerMiddleware } from 'botbuilder-applicationinsights';
import { IBotSettings } from '../services/botSettings.js';
import { ActivityEx, SkillsConfiguration } from 'bot-solutions/lib';
import { MainDialog } from '../dialogs/mainDialog';
import { Activity, BotFrameworkSkill, IEndOfConversationActivity } from 'botbuilder-core';

export class DefaultAdapter extends BotFrameworkAdapter {

    private readonly conversationState: ConversationState;
    private readonly telemetryClient: BotTelemetryClient;
    private readonly templateEngine: LocaleTemplateManager;
    private readonly skillClient: SkillHttpClient;
    private readonly skillsConfig: SkillsConfiguration;

    public constructor(
        settings: Partial<IBotSettings>,
        templateEngine: LocaleTemplateManager,
        conversationState: ConversationState,
        adapterSettings: Partial<BotFrameworkAdapterSettings>,
        telemetryMiddleware: TelemetryInitializerMiddleware,
        telemetryClient: BotTelemetryClient,
        skillsConfig: SkillsConfiguration,
        skillClient: SkillHttpClient
    ) {
        super(adapterSettings);

        if (conversationState === undefined) {
            throw new Error('conversationState parameter is null');
        }
        this.conversationState = conversationState;
        if (templateEngine === undefined) {
            throw new Error('templateEngine parameter is null');
        }
        this.templateEngine = templateEngine;
        if (telemetryClient === undefined) {
            throw new Error('telemetryClient parameter is null');
        }
        this.telemetryClient = telemetryClient;
        this.skillClient = skillClient;
        this.skillsConfig = skillsConfig;

        this.onTurnError = this.handleTurnError;

        if (settings.blobStorage === undefined) {
            throw new Error('There is no blobStorage value in appsettings file');
        }

        const transcriptStore: TranscriptStore = new BlobsTranscriptStore(settings.blobStorage.connectionString, settings.blobStorage.container);

        this.use(telemetryMiddleware);

        // Uncomment the following line for local development without Azure Storage
        // this.use(new TranscriptLoggerMiddleware(new MemoryTranscriptStore()));
        this.use(new TranscriptLoggerMiddleware(transcriptStore));
        this.use(new ShowTypingMiddleware());
        this.use(new SetLocaleMiddleware(settings.defaultLocale || 'en-us'));
        this.use(new EventDebuggerMiddleware());
        this.use(new SetSpeakMiddleware('en-US-JennyNeural', true));
    }

    private async handleTurnError(turnContext: TurnContext, error: Error): Promise<void> {
        // Log any leaked exception from the application.
        console.error(`[onTurnError] unhandled error : ${ error }`);

        await this.sendErrorMessage(turnContext, error);
        await this.endSkillConversation(turnContext);
        await this.clearConversationState(turnContext);
    }

    private async sendErrorMessage(turnContext: TurnContext, error: Error): Promise<void> {
        try {
            const telemetryException: TelemetryException = {
                exception: error
            };

            this.telemetryClient.trackException(telemetryException);

            // Send a message to the user.
            await turnContext.sendActivity(this.templateEngine.generateActivityForLocale('ErrorMessage', turnContext.activity.locale));

            // Send a trace activity, which will be displayed in the Bot Framework Emulator.
            // Note: we return the entire exception in the value property to help the developer;
            // this should not be done in production.
            await turnContext.sendTraceActivity('onTurnError Trace', error.message, 'https://www.botframework.com/schemas/error', 'TurnError');
        }
        catch (err) {
            console.error(`Exception caught in sendErrorMessage : ${ err }`);
        }
    }

    private async endSkillConversation(turnContext: TurnContext): Promise<void> {
        if (this.skillClient === undefined || this.skillsConfig === undefined) {
            return;
        }

        try {
            // Inform the active skill that the conversation is ended so that it has a chance to clean up.
            // Note: the root bot manages the ActiveSkillPropertyName, which has a value while the root bot
            // has an active conversation with a skill.
            const activeSkill: BotFrameworkSkill | undefined = await this.conversationState.createProperty<BotFrameworkSkill>(MainDialog.activeSkillPropertyName).get(turnContext);
            if (activeSkill !== undefined) {
                let endOfConversation: Partial<IEndOfConversationActivity> = ActivityEx.createEndOfConversationActivity();
                endOfConversation.code = 'RootSkillError';
                endOfConversation = TurnContext.applyConversationReference(endOfConversation, TurnContext.getConversationReference(turnContext.activity), true);

                await this.conversationState.saveChanges(turnContext, true);
                await this.skillClient.postToSkill(this.settings.appId, activeSkill, this.skillsConfig.skillHostEndpoint, endOfConversation as Activity);
            }
        }
        catch (err) {
            console.error(`Exception caught on attempting to send endOfConversation : ${ err }`);
        }
    }

    private async clearConversationState(turnContext: TurnContext): Promise<void> {
        try {
            // Delete the conversationState for the current conversation to prevent the
            // bot from getting stuck in a error-loop caused by being in a bad state.
            // ConversationState should be thought of as similar to "cookie-state" for a Web page.
            await this.conversationState.delete(turnContext);
        }
        catch (err) {
            console.error(`Exception caught on attempting to Delete ConversationState : ${ err }`);
        }
    }
}

File: samples/declarative/assistants/virtual-assistant/Adapters/DefaultAdapter.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Threading.Tasks;
using VirtualAssistantSample.Services;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Azure;
using Microsoft.Bot.Builder.Integration.AspNet.Core;
using Microsoft.Bot.Builder.TraceExtensions;
using Microsoft.Bot.Solutions.Middleware;
using Microsoft.Bot.Solutions.Responses;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace VirtualAssistantSample.Adapters
{
    public class DefaultAdapter : BotFrameworkHttpAdapter
    {
        private readonly BotSettings _settings;
        private readonly ILogger _logger;
        private readonly IStorage _storage;
        private readonly UserState _userState;
        private readonly ConversationState _conversationState;
        private readonly LocaleTemplateManager _templateEngine;
        private readonly IBotTelemetryClient _telemetryClient;

        public DefaultAdapter(
            IConfiguration configuration, 
            BotSettings settings,
            ILogger<BotFrameworkHttpAdapter> logger, 
            IStorage storage, 
            UserState userState, 
            ConversationState conversationState,
            LocaleTemplateManager templateEngine,
            IBotTelemetryClient telemetryClient)
            : base(configuration, logger: logger)
        {
            _settings = settings ?? throw new ArgumentNullException(nameof(settings));
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
            _storage = storage ?? throw new ArgumentNullException(nameof(storage));

            _userState = userState ?? throw new ArgumentNullException(nameof(userState));
            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));

            _templateEngine = templateEngine ?? throw new ArgumentNullException(nameof(templateEngine));
            _telemetryClient = telemetryClient ?? throw new ArgumentNullException(nameof(telemetryClient));

            OnTurnError = HandleTurnErrorAsync;

            this.UseStorage(_storage);
            this.UseState(_userState, _conversationState);

            Use(new TranscriptLoggerMiddleware(new AzureBlobTranscriptStore(_settings.BlobStorage.ConnectionString, _settings.BlobStorage.Container)));
            Use(new ShowTypingMiddleware());
            Use(new SetLocaleMiddleware(_settings.DefaultLocale ?? "en-us"));
            Use(new EventDebuggerMiddleware());
            Use(new SetSpeakMiddleware());
        }

        private async Task HandleTurnErrorAsync(ITurnContext turnContext, Exception exception)
        {
            // Log any leaked exception from the application.
            _logger.LogError(exception, $"[OnTurnError] unhandled error : {exception.Message}");

            await SendErrorMessageAsync(turnContext, exception);
            // TODO: Enable Skills
            // await EndSkillConversationAsync(turnContext);
            await ClearConversationStateAsync(turnContext);
        }

        private async Task SendErrorMessageAsync(ITurnContext turnContext, Exception exception)
        {
            try
            {
                _telemetryClient.TrackException(exception);

                // Send a message to the user.
                await turnContext.SendActivityAsync(_templateEngine.GenerateActivityForLocale("ErrorMessage"));

                // Send a trace activity, which will be displayed in the Bot Framework Emulator.
                // Note: we return the entire exception in the value property to help the developer;
                // this should not be done in production.
                await turnContext.TraceActivityAsync("OnTurnError Trace", exception.ToString(), "https://www.botframework.com/schemas/error", "TurnError");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Exception caught in SendErrorMessageAsync : {ex}");
            }
        }

        // TODO: Enable Skills
        //private async Task EndSkillConversationAsync(ITurnContext turnContext)
        //{
        //    if (_skillClient == null || _skillsConfig == null)
        //    {
        //        return;
        //    }

        //    try
        //    {
        //        // Inform the active skill that the conversation is ended so that it has a chance to clean up.
        //        // Note: the root bot manages the ActiveSkillPropertyName, which has a value while the root bot
        //        // has an active conversation with a skill.
        //        // TODO: Determine where to handle Active Skill name
        //        var activeSkill = await _conversationState.CreateProperty<BotFrameworkSkill>(DispatchDialog.ActiveSkillPropertyName).GetAsync(turnContext, () => null);
        //        if (activeSkill != null)
        //        {
        //            var endOfConversation = Activity.CreateEndOfConversationActivity();
        //            endOfConversation.Code = "RootSkillError";
        //            endOfConversation.ApplyConversationReference(turnContext.Activity.GetConversationReference(), true);

        //            await _conversationState.SaveChangesAsync(turnContext, true);
        //            await _skillClient.PostActivityAsync(_settings.MicrosoftAppId, activeSkill, _skillsConfig.SkillHostEndpoint, (Activity)endOfConversation, CancellationToken.None);
        //        }
        //    }
        //    catch (Exception ex)
        //    {
        //        _logger.LogError(ex, $"Exception caught on attempting to send EndOfConversation : {ex}");
        //    }
        //}

        private async Task ClearConversationStateAsync(ITurnContext turnContext)
        {
            try
            {
                // Delete the conversationState for the current conversation to prevent the
                // bot from getting stuck in a error-loop caused by being in a bad state.
                // ConversationState should be thought of as similar to "cookie-state" for a Web page.
                await _conversationState.DeleteAsync(turnContext);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Exception caught on attempting to Delete ConversationState : {ex}");
            }
        }
    }
}
